Fix URL creation for ftp_site field in dataset table when using files metadata console tool

pli888 opened the pull request:
# Pull request for issue: #1466

This is a pull request for the following functionalities:

* A fix for a bug when transforming a dataset's ftp_site value that contains two `/pub/` substrings, e.g. `https://ftp.cngb.org/pub/gigadb/pub/10.5524/100001_101000/100003/` would result in `https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/staging/gigadb` which is not correct.

## How to test?

### Dev environment

Spin up a local GigaDB instance:
```
$ ./up.sh
```

Change directory:
```
$ cd gigadb/app/tools/files-metadata-console
```

If you look in the dataset table, you will see that dataset 100935 will have a ftp_site value of `https://ftp.cngb.org/pub/gigadb/pub/10.5524/100001_101000/100935/` which contains two `/pub/` sub-strings.

Transform all dataset ftp_site and file location URLs using files metadata console tool:
```
$ docker-compose run --rm files-metadata-console ./yii update/urls --separator=/pub/ --next=10 --apply
```

If you now check the ftp_site field for dataset 100935, you should see that it is `https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/dev/pub/10.5524/100001_101000/100935/` which is now correct.

### Staging environment

Deploy the code in this PR onto your staging server from your CI/CD pipeline.

Log into your bastion server.

Drop datasbase triggers:
```
[centos@ip-10-99-0-27 ~]$ $ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c 'drop trigger if exists file_finder_trigger on file RESTRICT'
-bash: $: command not found
[centos@ip-10-99-0-27 ~]$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c 'drop trigger if exists file_finder_trigger on file RESTRICT'
DROP TRIGGER
[centos@ip-10-99-0-27 ~]$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c 'drop trigger if exists sample_finder_trigger on sample RESTRICT'
DROP TRIGGER
[centos@ip-10-99-0-27 ~]$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c 'drop trigger if exists dataset_finder_trigger on dataset RESTRICT'
DROP TRIGGER
```

Transform a couple of dataset ftp_site and file location URLs using files metadata console tool:
```
[centos@ip-10-99-0-27 ~]$ docker run --rm "registry.gitlab.com/$GITLAB_PROJECT/production-files-metadata-console:$GIGADB_ENV" ./yii update/urls --separator=/pub/ --stop=102443 --next=2 --apply
        Transforming ftp_site for dataset 100001... 
DONE

        Transforming file locations for dataset 100001...
DONE (19/19)

        Transforming ftp_site for dataset 100002... 
DONE

        Transforming file locations for dataset 100002...
DONE (10/10)
```

If you check the ftp_site field for dataset 100001, you should see this value: `https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/staging/pub/10.5524/100001_101000/100001`

## How have functionalities been implemented?

The bug was caused by the previous functionality in the DatasetFilesURLUpdater.php component class not taking into account for when a ftp_site URL contained two `/pub/` substrings. This would lead to the second string token: `gigadb` being selected and concatenated into the ftp_site URL under creation. The new functionality now selects the last token which will look something like `10.5524/100001_101000/100245/`. The last token will be added to the end of `https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/$DEPLOYMENT_ENV/pub/` to form the correct ftp_site URL.

## Any issues with implementation?

As suggested by @rija, the `WARN[0000] a network with name deployment_db-tier exists but was not created for project "files-metadata-console".` error message can be resolved by creating a `docker-compose.override.yml` in gigadb/app/tools/files-metadata-console with content:
```
version: '3.7'
networks:
  proxy-db-tier:
    name: deployment_db-tier
    external: true
```

## Any changes to automated tests?

The `tryReplaceFleUrlSubstringWithPrefix()` functional test in `ReplaceFileUrlSubstringWithPrefixCest` has been updated to test for this bug using dataset 100005 in the gigadb_testdata database.

pli888 committed:
Add file metadata for dataset 100935 in dev data

Data for a file belonging to dataset 100935 has been added to
replicate bug #1466.

pli888 committed:
Fix bug when ftp_site value contains two pub substrings

Functional test has been updated to check for this bug.

pli888 committed:
Add entry into changelog

rija reviewed (approved):
Hi @pli888,

**praise**: following the testing instructions I was able to get the bug fix working locally and on my staging
**praise**: all tests are passing
**issue** (non-blocking): there isn't any consequence in this PR on forks pipeline, but never DROP triggers without recreating them afterward, this break the search functionality (something to bear in mind when doing the real url replacement)
**note**: I found a better workaround for the existing network error on local deployment:
Instead of messing with the git versioned docker-compose file, create a `docker-compose.override.yml` in `gigadb/app/tools/files-metadata-console` with content: 
```
version: '3.7'
networks:
  proxy-db-tier:
    name: deployment_db-tier
    external: true
```
And that's it, it will just work with the exact same docker-compose command and arguments.
Ideally that new file should be git ignored, as it's only needed locally, and we don't want to accidentally commit it
**suggestion** (non-blocking): add `docker-compose.override.yml` to the local project's `.gitignore`
**praise**: code looks fine

Based on the above, I'm happy to approve.

pli888 committed:
Update .gitignore

Do not commit local docker-compose.override.yml file which fixes
existing network error.

pli888 commented:
>suggestion (non-blocking): add docker-compose.override.yml to the local project's .gitignore

This has been implemented now in this [commit](https://github.com/gigascience/gigadb-website/pull/1483/commits/271575a5d495e0ea73e1f5788939b1caf9b1e360).

>issue (non-blocking): there isn't any consequence in this PR on forks pipeline, but never DROP triggers without recreating them afterward, this break the search functionality (something to bear in mind when doing the real url replacement)

Yes, there is the upcoming PR #1448 which will update documentation for using the file location update tool in the files metadata console. There is a `Staging and live environments` section in the documentation which contains steps to re-create triggers.

kencho51 reviewed (approved):
Hi @pli888,

**praise:** The url in `ftp_site` of table `dataset` and `location` of table `file` shows the correct path, not truncated.
**praise:** The updated functional test `tryReplaceFileUrlSubstringWithPrefix` makes sure the tool working as expected.
**issue(non-blocking):** The prefix `https://ftp.cngb.org/pub/gigadb/` can be found in `tests/acceptance/SearchAuthors.feature`, `tests/acceptance/Search.feature` and `tests/acceptance/AdminFileForm.feature`, these tests are still passing locally and in gitlab pipeline. The reasons for this are:
- locally, the test database would be restored before each acceptance test,whereas the test data contains `https://ftp.cngb.org/pub/gigadb/`.
- in gitlab, files meta console tool has not been executed, so the `ftp_site` and `location` have not been updated yet.

So the acceptance tests seem not capturing the scenarios quite right, and the tests  result cannot show that.
However, the functional test `tryReplaceFileUrlSubstringWithPrefix` has confirmed the tool can work correctly by asserting the `ftp_site` and `location` have been updated correctly, therefore, from the perspective of the tool itself, I don't think it should be blocked from getting merged.
**suggestion(non-blocking):** The test data for the test database need to be updated `go live`. And I have reported this issue at https://github.com/gigascience/gigadb-website/issues/1488

Code LGTM, happy to approve.
